### DISCUSSION TITLE
How i got GPU / CUDA working (Guide)

### DISCUSSION THREAD

**Discussion by mrdarj** [imartinez/privateGPT#928] [General]
This is how i got GPU support working, as a note i am using venv within PyCharm in Windows 11

Compute time is down to around 15 seconds on my 3070 Ti using the included txt file, some tweaking will likely speed this up

Prerequisite is to have CUDA Drivers installed, in my case NVIDIA CUDA Drivers

1 - We need to remove Llama and reinstall version with CUDA support, so:
` pip uninstall llama-cpp-python   `

2 - We need to find the correct version of llama to install, we need to know:

a) Installed CUDA version, type `nvidia-smi` inside PyCharm or Windows Powershell, shows CUDA version eg 12.2 

b) CPU AVX support, google it but an easy way is if you have Steam to go help > System Information and check which AVX is supported, eg AVX2

c) current Python version eg 3.11

3 - Now we know we need (eg) AVX2, CUDA 12.2, py3.11 Llama, go here for Llama versions:

https://[jllllll.github.io/llama-cpp-python-cuBLAS-wheels/](https://jllllll.github.io/llama-cpp-python-cuBLAS-wheels/)

and go here for more info:

https://github.com/jllllll/llama-cpp-python-cuBLAS-wheels

4 - Find within the directories the correct version, so i went AVX2 > CUDA 12.2 > llama_cpp_python > and my file is called llama_cpp_python-0.1.77+cu122-cp311-cp311-win_amd64.whl - cu122 (CUDA), cp311 (python)

5 - Right click and copy link to this correct llama version

6 - Inside PyCharm, 

`pip install **Link** `

This installed llama-cpp-python with CUDA support directly from the link we found above

7 - Inside privateGPT.py:

add `model_n_gpu = os.environ.get('MODEL_N_GPU')`

This is just a custom variable for GPU offload layers

change            ` llm = LlamaCpp(model_path=model_path, n_ctx=model_n_ctx, max_tokens=model_n_ctx, n_gpu_layers=model_n_gpu, n_batch=model_n_batch, callbacks=callbacks, verbose=False)`

We add the GPU offload settings, and we add n_ctx which is the chunk size context window (from https://github.com/abetlen/llama-cpp-python)

8 - Inside .env:

```
MODEL_TYPE=LlamaCpp
MODEL_PATH=models/wizardLM-7B.ggmlv3.q4_0.bin
EMBEDDINGS_MODEL_NAME=all-MiniLM-L6-v2
MODEL_N_CTX=1792
MODEL_N_BATCH=8
TARGET_SOURCE_CHUNKS=4
MODEL_N_GPU=40
```

We adjust the model type to llama, the model to a specifically chosen one, the CTX, the batch, and the GPU layers

9 - Download the Model (you can use any that work with llama)

https://huggingface.co/TheBloke/wizardLM-7B-GGML

10 - Put model into models folder in project

11 - Run project (privateGPT.py)

If CUDA is working you should see this as the first line of the program:

```
ggml_init_cublas: found 1 CUDA devices:
Device 0: NVIDIA GeForce RTX 3070 Ti, compute capability 8.6
```

(With your model GPU)

You should see` llama_model_load_internal: n_ctx      = 1792`

If this is 512 you will likely run out of token size from a simple query

You should see `llama_model_load_internal: offloaded 35/35 layers to GPU`

This is the amount of layers we offload to GPU (As our setting was 40)

You can set this to 20 as well to spread load a bit between GPU/CPU, or adjust based on your specs

You should see GPU usage high when running queries

If the above works then you should have full CUDA / GPU support

It just comes down to reinstalling llama, changing a few settings, and making sure we have a working model

Enjoy

**Comment by kon75** [comment]
Hello mrdarj, I'm super interested in testing out your modifications.

I ended uninstalling pip uninstall llama-cpp-python at the very end again 
and installing pip install llama-cpp-python==0.1.78
due to the following suggestion I found...
[(https://github.com/langchain-ai/langchain/issues/2485)](https://user-images.githubusercontent.com/16869789/265252594-2104e824-7f43-430d-9153-10b0129785cc.png)

now the response time appears to be pretty good, around 4-8 sec on easy queries.

**Comment by sp1d3rino** [comment]
hi, I'm using conda to install privateGPT.
I followed your indications but I get this error

File "pydantic\main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for LlamaCpp
__root__
  Could not load Llama model from path: D:/models/wizardLM-7B.ggmlv3.q4_0.bin. Received error Failed to load shared library 'C:\Users\..\anaconda3\envs\privateGPT\lib\site-packages\llama_cpp\llama.dll': Could not find module 'C:\Users\..\anaconda3\envs\privateGPT\lib\site-packages\llama_cpp\llama.dll' (or one of its dependencies). Try using the full path with constructor syntax. (type=value_error)

how can I fix it?

**Comment by Hempstead** [comment]
Hi, I am not able to load wizardLM-7B.ggmlv3.q4_0.bin with your setup. I am on SUSE linux (15.SP4).

My ggml-gpt4all-j-v1.3-groovy.bin works fine.

Any suggestions?

**Reply by levoniust** [reply]
Not real help sorry, but you are trying to use a ggml that runs on cpu, you need a gptq model because the above prosses makes the model run on gpu.
